src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/RoleSyncService.kt

```diff
@@ -62,33 +63,44 @@ class RoleSyncService(
     return filter { r -> rolesMap[r.key] == null }
   }
 
-  fun syncRole(currentRoleData: RoleDataToSync?, newRoleData: RoleDataToSync?, stats: SyncStatistics) {
+  fun syncRole(currentRoleData: RoleDataToSync?, newRoleData: RoleDataToSync?, stats: SyncStatistics, readOnly: Boolean) {
 
     val diff = checkForDifferences(currentRoleData, newRoleData)
-    if (diff.entriesOnlyOnLeft().isNotEmpty()) {
-      stats.roles[currentRoleData!!.roleCode] = RoleDifferences(currentRoleData.roleCode, diff.toString())
-      log.error("Role exists but should be deleted {}", currentRoleData.roleCode)
-      telemetryClient.trackEvent("HMUA-Role-Change-Failure", mapOf("roleCode" to currentRoleData.roleCode), null)
-    } else if (!diff.areEqual()) {
-      stats.roles[newRoleData!!.roleCode] = RoleDifferences(newRoleData.roleCode, diff.toString())
-      try {
-        storeInNomis(currentRoleData, newRoleData, stats)
-        if (stats.roles[newRoleData.roleCode]?.updateType != NONE) {
-          val trackingAttributes = mapOf(
-            "roleCode" to newRoleData.roleCode,
-            "differences" to stats.roles[newRoleData.roleCode]?.differences,
-          )
-          telemetryClient.trackEvent("HMUA-Role-Change", trackingAttributes, null)
-        }
-      } catch (e: Exception) {
-        stats.roles[newRoleData.roleCode] = stats.roles[newRoleData.roleCode]!!.copy(updateType = ERROR)
 
-        log.error("Failed to update {} - message = {}", newRoleData.roleCode, e.message)
-        telemetryClient.trackEvent("HMUA-Role-Change-Failure", mapOf("roleCode" to newRoleData.roleCode), null)
+    if (!diff.areEqual()) {
+      diff.storeStatistics(currentRoleData, newRoleData, stats)
+      if (readOnly) return
+
```

Comment:
@petergphillips now store the statistics first to enable a neater return if readonly.  It does mean we end up doing two calls to diff.entriesOnlyOnLeft, but is more readable.